File: packages/agent/src/subagent/commands.ts
// The `demi agent` command-tree definition. Pure declaration: every action
// goes through `SubagentCommandOps`; the supervisor owns child lifecycles.
import { createId, errorMessage } from '@demicodes/utils'
import { z } from 'zod'
import {
  isCommandGroup,
  type Command,
  type CommandGroup,
  type CommandStorage
} from '@demicodes/shell'
import { flattenTree, renderTreeNode, type AgentTreeNode } from './format'

const requestIdSchema = z.string().min(1).max(128).optional().describe(
  'Stable id for this creation or resume request. Supply the same id and arguments to retry safely after an uncertain response; otherwise a new id is generated.'
)

const SPAWN_PROMPT_DESCRIPTION =
  "The child's first user message and only task brief. The child starts with an empty transcript and cannot see this conversation: do not refer to prior turns, and do not paste this conversation or the product user's message unchanged. Include the goal for this child, applicable decisions and constraints, whether to edit or only report, how to verify, and every concrete identifier it needs (paths, ids, error text, commands already tried and their key results). State the exact shape of the last assistant text it should return."

/** The lifecycle, communication, and read operations the command tree needs. */
export interface SubagentCommandOps {
  canSpawn: boolean
  profileNames(): string[]
  spawn(input: {
    prompt: string;
    profileName: string | undefined;
    description: string;
    isSpawnForbidden: boolean
  }, requestId: string, storage: CommandStorage): Promise<string>
  resumeArchived(id: string, message: string, requestId: string, storage: CommandStorage): Promise<string>
  getRunning(id: string): unknown | null
  send(id: string, message: string): Promise<string>
  abortSubtree(id: string): Promise<void>
  tree(): Promise<AgentTreeNode[]>
  ownerId(): string
  show(
    id: string
  ): {
    snapshot: Record<string, unknown>;
    text: string
  } | null
}

export function subagentCommandNode(
  ops: SubagentCommandOps
): CommandGroup {
  const profileNames = ops.profileNames()
  const subcommands: Command[] = [
    {
      name: 'spawn',
      kind: 'rpc',
      summary:
        'Start a child agent session and return its id immediately after creation. The child runs independently of this command. Completion arrives as a message to the parent, waking it when idle. When you have no independent work left, end your turn and let the completion message wake you. Do not poll agent list/show or schedule timed yield calls to wait for children. Use agent send to communicate and agent abort to stop it. Children can spawn children of their own.',
      successOutput:
        'stdout is "subagentId: <id>"; creation succeeded, not necessarily execution',
      failureOutput: 'non-zero exit with the creation failure reason on stderr',
      input: {
        prompt: z.string().describe(SPAWN_PROMPT_DESCRIPTION),
        'request-id': requestIdSchema,
        profile: z
          .string()
          .optional()
          .describe(`Named subagent profile configured at harness assembly; omit to inherit the parent's model, prompt, Host and commands. Available: ${profileNames.length > 0 ? profileNames.join(', ') : 'none'}.`),
        description: z
          .string()
          .optional()
          .describe('Short UI title distinguishing concurrent children.'),
        'no-subagents': z.boolean().optional().describe(
          'Forbid this child from spawning subagents of its own; it can still send, list, and show.'
        ),
      },
      stdinField: 'prompt',
      output: { json: z.object({ subagentId: z.string() }) },
      run: async ({ parsed, io, storage }) => {
        const prompt = (parsed.values.prompt as string).trim()
        if (!prompt) {
          await io.stderr('demi agent spawn: prompt must not be empty\n')
          return { exitCode: 1 }
        }
        let subagentId: string
        try {
          subagentId = await ops.spawn({
            prompt,
            profileName: parsed.values.profile === undefined
              ? undefined
              : String(parsed.values.profile),
            description: parsed.values.description === undefined
              ? ''
              : String(parsed.values.description),
            isSpawnForbidden: parsed.values['no-subagents'] === true,
          }, String(parsed.values['request-id'] ?? createId()), storage)
        } catch (error) {
          await io.stderr(`demi agent spawn: ${errorMessage(error)}\n`)
          return { exitCode: 1 }
        }
        await io.stdout(parsed.json
          ? `${JSON.stringify({ subagentId })}\n`
          : `subagentId: ${subagentId}\n`)
        return { exitCode: 0 }
      },
    },
    {
      name: 'send',
      summary:
        'Deliver information to any live agent in the tree, or parent. A busy recipient incorporates it through internal steering; an idle recipient wakes. Returns after durable acceptance, without waiting for an answer. Use for interim information, questions, or blockers; your final answer is delivered automatically. Archived recipients must be reopened by their parent with resume.',
      input: {
        id: z.string().describe(
          'Target agent id from the tree, or "parent" for the session that spawned this one'
        ),
        message: z.string()
          .describe('Message body.'),
      },
      positionals: ['id'],
      stdinField: 'message',
      output: { json: z.object({ id: z.string(), accepted: z.boolean() }) },
      kind: 'rpc',
      run: async ({ parsed, io }) => {
        const message = (parsed.values.message as string).trim()
        if (!message) {
          await io.stderr('demi agent send: message must not be empty\n')
          return { exitCode: 1 }
        }
        try {
          const targetId = await ops.send(String(parsed.values.id), message)
          await io.stdout(parsed.json
            ? `${JSON.stringify({ id: targetId, accepted: true })}\n`
            : `sent to ${targetId}\n`)
          return { exitCode: 0 }
        } catch (error) {
          await io.stderr(`demi agent send: ${errorMessage(error)}\n`)
          return { exitCode: 1 }
        }
      },
    },
    {
      name: 'abort',
      summary: 'Abort one of your own running children and its whole subtree. Siblings are untouched; only the spawning session may abort a child.',
      input: { id: z.string().describe('subagentId from spawn stdout') },
      positionals: ['id'],
      output: { json: z.object({ id: z.string(), aborted: z.boolean() }) },
      kind: 'rpc',
      run: async ({ parsed, io }) => {
        const id = String(parsed.values.id)
        if (!ops.getRunning(id)) {
          await io.stderr(
            `demi agent abort: "${id}" is not one of your running children\n`
          )
          return { exitCode: 1 }
        }
        await ops.abortSubtree(id)
        await io.stdout(parsed.json
          ? `${JSON.stringify({ id, aborted: true })}\n`
          : `aborted ${id}\n`)
        return { exitCode: 0 }
      },
    },
    {
      name: 'resume',
      summary:
        'Revive one of your own archived children with a new user message on its preserved transcript. Return its id immediately after accepting the message; completion is delivered separately to the parent. Use agent send to communicate and agent abort to stop it. Archived ids are in agent list.',
      input: {
        id: z.string().describe('subagentId of an archived child'),
        'request-id': requestIdSchema,
        message: z.string().describe(
          'The reviving user message.'
        ),
      },
      positionals: ['id'],
      stdinField: 'message',
      output: { json: z.object({ subagentId: z.string() }) },
      kind: 'rpc',
      run: async ({ parsed, io, storage }) => {
        const id = String(parsed.values.id)
        const message = (parsed.values.message as string).trim()
        if (!message) {
          await io.stderr('demi agent resume: message must not be empty\n')
          return { exitCode: 1 }
        }
        let subagentId: string
        try {
          subagentId = await ops.resumeArchived(id, message, String(parsed.values['request-id'] ?? createId()), storage)
        } catch (error) {
          await io.stderr(`demi agent resume: ${errorMessage(error)}\n`)
          return { exitCode: 1 }
        }
        await io.stdout(parsed.json
          ? `${JSON.stringify({ subagentId })}\n`
          : `subagentId: ${subagentId}\n`)
        return { exitCode: 0 }
      },
    },
    {
      name: 'list',
      summary:
        'Render the whole session tree from the root down, marking your own position. Live agents show phase, ages, execution, and activity; each node\'s archived (finished, revivable by its parent) children render beneath it. Every age is relative to now. A read, not a wait — not for polling loops.',
      output: { json: z.object({ tree: z.array(z.unknown()) }) },
      kind: 'rpc',
      run: async ({ parsed, io }) => {
        const nodes = await ops.tree()
        if (parsed.json) {
          await io.stdout(
            `${JSON.stringify({ tree: flattenTree(nodes, ops.ownerId()) })}\n`
          )
          return { exitCode: 0 }
        }
        const lines: string[] = []
        for (const node of nodes) renderTreeNode(
          node,
          '',
          true,
          ops.ownerId(),
          lines
        )
        await io.stdout(`${lines.join('\n')}\n`)
        return { exitCode: 0 }
      },
    },
    {
      name: 'show',
      summary:
        'Bounded snapshot of any live agent in the tree (root excluded): execution state, recent tool titles with durations, last assistant text. Every duration is relative to now — use the ages to tell motion from stall. Omits tool outputs, file contents, and older turns. A read, not a wait — not for polling loops.',
      input: { id: z.string().describe('Agent id from the tree') },
      positionals: ['id'],
      output: { json: z.object({ agent: z.unknown() }) },
      kind: 'rpc',
      run: async ({ parsed, io }) => {
        const id = String(parsed.values.id)
        const entry = ops.show(id)
        if (!entry) {
          await io.stderr(`demi agent show: no live agent "${id}"\n`)
          return { exitCode: 1 }
        }
        if (parsed.json) await io.stdout(`${JSON.stringify({ agent: entry.snapshot })}\n`)
        else await io.stdout(entry.text)
        return { exitCode: 0 }
      },
    },
  ]
  return {
    name: 'agent',
    summary: ops.canSpawn
      ? 'Agent tree: spawn and manage your own children; send, list and show any live agent.'
      : 'Agent tree communication: this session may not spawn subagents; send, list and show any live agent.',
    subcommands: subcommands.filter((command) => ops.canSpawn || ![
      'spawn',
      'abort',
      'resume'
    ].includes(command.name)),
  }
}

/**
 * The `agent` node's shape alone, for a manifest built outside any session:
 * a target parses `demi agent …` from it and relays the call as `rpc`, which
 * runs against the live session's tree. Running the shape itself is a wiring
 * error.
 */
export function subagentCommandShape(profileNames: string[]): CommandGroup {
  const notHere = (): never => {
    throw new Error(
      'demi agent runs on the live session; the manifest carries its shape only'
    )
  }
  return subagentCommandNode({
    canSpawn: true,
    profileNames: () => profileNames,
    spawn: notHere,
    resumeArchived: notHere,
    getRunning: notHere,
    send: notHere,
    abortSubtree: notHere,
    tree: notHere,
    ownerId: notHere,
    show: notHere,
  })
}

/**
 * Grafts the `agent` node under a `demi` root: onto an existing harness `demi`
 * tree, or as a new `demi` root when the harness has none.
 */
export function injectSubagentCommand(
  commands: Command[],
  agentNode: CommandGroup
): Command[] {
  const demiIndex = commands.findIndex((command) => command.name === 'demi')
  if (demiIndex === -1) {
    return [
      ...commands,
      {
        name: 'demi',
        summary: 'Demi agent runtime commands.',
        subcommands: [agentNode]
      }
    ]
  }
  const demi = commands[demiIndex]!
  if (!isCommandGroup(demi))
    throw new Error('injectSubagentCommand: the demi root must be a group')
  const subcommands = [
    ...demi.subcommands.filter((command) => command.name !== 'agent'),
    agentNode
  ]
  const next = [...commands]
  next[demiIndex] = { ...demi, subcommands }
  return next
}

File: crates/agent/src/server/commands.rs
//! The `demi agent` command group (`subagents.md` § Model-facing surface):
//! its declarations, and the handlers that run each call against the live
//! tree of the invoking job's conversation. Lifecycle (`spawn`, `abort`,
//! `resume`) acts on the caller's own children; communication and reads
//! reach any live agent of the tree.

use std::rc::{Rc, Weak};

use demi_core::{NodeId, is_blank, trim};
use demi_shell::{
    Call, CommandSet, GroupBuilder, LeafBuilder, RegisterError, RpcError, RpcPort, TypedRpc,
};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};

use super::{
    AgentServer,
    tree::{AgentSnapshot, StartInput, Tree, TreeEntry},
};
use crate::{AgentHarness, Profile};

/// The one description of a spawn's brief (`subagents.md` § Command help).
const SPAWN_PROMPT: &str = "The child's first user message and only task brief. The child starts with an empty transcript and cannot see this conversation: do not refer to prior turns, and do not paste this conversation or the product user's message unchanged. Include the goal for this child, applicable decisions and constraints, whether to edit or only report, how to verify, and every concrete identifier it needs (paths, ids, error text, commands already tried and their key results). State the exact shape of the last assistant text it should return.";

const SPAWN_SUMMARY: &str = "Start a child agent session and return its id immediately after creation. The child runs independently of this command. Completion arrives as a message to the parent, waking it when idle. When you have no independent work left, end your turn and let the completion message wake you. Do not poll agent list/show or schedule timed yield calls to wait for children. Use agent send to communicate and agent abort to stop it. Children can spawn children of their own.";

const SEND_SUMMARY: &str = "Deliver information to any live agent in the tree, or parent. A busy recipient incorporates it through internal steering; an idle recipient wakes. Returns after durable acceptance, without waiting for an answer. Use for interim information, questions, or blockers; your final answer is delivered automatically. Archived recipients must be reopened by their parent with resume.";

const ABORT_SUMMARY: &str = "Abort one of your own running children and its whole subtree. Siblings are untouched; only the spawning session may abort a child.";

const RESUME_SUMMARY: &str = "Revive one of your own archived children with a new user message on its preserved transcript. Return its id immediately after accepting the message; completion is delivered separately to the parent. Use agent send to communicate and agent abort to stop it. Archived ids are in agent list.";

const LIST_SUMMARY: &str = "Render the whole session tree from the root down, marking your own position. Live agents show phase, ages, execution, and activity; each node's archived (finished, revivable by its parent) children render beneath it. Every age is relative to now. A read, not a wait — not for polling loops.";

const SHOW_SUMMARY: &str = "Bounded snapshot of any live agent in the tree (root excluded): execution state, recent tool titles with durations, last assistant text. Every duration is relative to now — use the ages to tell motion from stall. Omits tool outputs, file contents, and older turns. A read, not a wait — not for polling loops.";

/// The input of `demi agent spawn`.
#[derive(Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
struct SpawnArgs {
    prompt: String,
    /// Stable id for this creation or resume request. Supply the same id and
    /// arguments to retry safely after an uncertain response; otherwise a new
    /// id is generated.
    #[schemars(length(min = 1, max = 128))]
    request_id: Option<String>,
    profile: Option<String>,
    /// Short UI title distinguishing concurrent children.
    description: Option<String>,
    /// Forbid this child from spawning subagents of its own; it can still
    /// send, list, and show.
    no_subagents: Option<bool>,
}

/// The input of `demi agent send`.
#[derive(Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
struct SendArgs {
    /// Target agent id from the tree, or "parent" for the session that
    /// spawned this one
    id: String,
    /// Message body.
    message: String,
}

/// The input of `demi agent abort`.
#[derive(Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
struct AbortArgs {
    /// subagentId from spawn stdout
    id: String,
}

/// The input of `demi agent resume`.
#[derive(Deserialize, JsonSchema)]
#[serde(deny_unknown_fields, rename_all = "kebab-case")]
struct ResumeArgs {
    /// subagentId of an archived child
    id: String,
    /// Stable id for this creation or resume request. Supply the same id and
    /// arguments to retry safely after an uncertain response; otherwise a new
    /// id is generated.
    #[schemars(length(min = 1, max = 128))]
    request_id: Option<String>,
    /// The reviving user message.
    message: String,
}

/// The input of `demi agent list`, which takes none.
#[derive(Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
struct ListArgs {}

/// The input of `demi agent show`.
#[derive(Deserialize, JsonSchema)]
#[serde(deny_unknown_fields)]
struct ShowArgs {
    /// Agent id from the tree
    id: String,
}

/// `demi agent spawn --json` and `resume --json`.
#[derive(Serialize, JsonSchema)]
#[serde(rename_all = "camelCase")]
struct Started {
    subagent_id: NodeId,
}

/// `demi agent send --json`.
#[derive(Serialize, JsonSchema)]
struct Sent {
    id: NodeId,
    accepted: bool,
}

/// `demi agent abort --json`.
#[derive(Serialize, JsonSchema)]
struct Aborted {
    id: NodeId,
    aborted: bool,
}

/// `demi agent list --json`.
#[derive(Serialize, JsonSchema)]
struct Listing {
    tree: Vec<TreeEntry>,
}

/// `demi agent show --json`.
#[derive(Serialize, JsonSchema)]
struct Shown {
    agent: AgentSnapshot,
}

/// A node's commands: its harness's, with the `demi agent` group grafted
/// under a `demi` group, replacing any `agent` child the harness declared,
/// or under a new `demi` root.
pub(crate) fn with_agent_group<H: AgentHarness>(
    harness_commands: &CommandSet,
    server: &Rc<AgentServer<H>>,
    can_spawn: bool,
    profiles: &[Profile],
) -> Result<CommandSet, RegisterError> {
    let mut commands = harness_commands.clone();
    let group = agent_group(Rc::downgrade(server), can_spawn, profiles);
    let has_demi = commands.declarations().any(|root| root.name() == "demi");
    if has_demi {
        commands.graft(&["demi"], group)?;
    } else {
        commands
            .register(GroupBuilder::new("demi", "Demi agent runtime commands.").group(group))?;
    }
    Ok(commands)
}

/// The `agent` group of a node: every verb, or communication and reads only
/// when the node may not spawn.
fn agent_group<H: AgentHarness>(
    server: Weak<AgentServer<H>>,
    can_spawn: bool,
    profiles: &[Profile],
) -> GroupBuilder {
    let summary = if can_spawn {
        "Agent tree: spawn and manage your own children; send, list and show any live agent."
    } else {
        "Agent tree communication: this session may not spawn subagents; send, list and show any live agent."
    };
    let names: Vec<&str> = profiles
        .iter()
        .map(|profile| profile.name.as_str())
        .collect();
    let available = if names.is_empty() {
        "none".to_owned()
    } else {
        names.join(", ")
    };
    let mut group = GroupBuilder::new("agent", summary);
    if can_spawn {
        group = group.leaf(
            LeafBuilder::rpc("spawn", SPAWN_SUMMARY)
                .input::<SpawnArgs>()
                .describe("prompt", SPAWN_PROMPT)
                .describe(
                    "profile",
                    format!("Named subagent profile configured at harness assembly; omit to inherit the parent's model, prompt, Host and commands. Available: {available}."),
                )
                .stdin_field("prompt")
                .success_output("stdout is \"subagentId: <id>\"; creation succeeded, not necessarily execution")
                .failure_output("non-zero exit with the creation failure reason on stderr")
                .json_output::<Started>()
                .bind(TypedRpc::new(verb(server.clone(), spawn))),
        );
    }
    group = group.leaf(
        LeafBuilder::rpc("send", SEND_SUMMARY)
            .input::<SendArgs>()
            .positionals(["id"])
            .stdin_field("message")
            .json_output::<Sent>()
            .bind(TypedRpc::new(verb(server.clone(), send))),
    );
    if can_spawn {
        group = group
            .leaf(
                LeafBuilder::rpc("abort", ABORT_SUMMARY)
                    .input::<AbortArgs>()
                    .positionals(["id"])
                    .json_output::<Aborted>()
                    .bind(TypedRpc::new(verb(server.clone(), abort))),
            )
            .leaf(
                LeafBuilder::rpc("resume", RESUME_SUMMARY)
                    .input::<ResumeArgs>()
                    .positionals(["id"])
                    .stdin_field("message")
                    .json_output::<Started>()
                    .bind(TypedRpc::new(verb(server.clone(), resume))),
            );
    }
    group
        .leaf(
            LeafBuilder::rpc("list", LIST_SUMMARY)
                .json_output::<Listing>()
                .bind(TypedRpc::new(verb(server.clone(), list))),
        )
        .leaf(
            LeafBuilder::rpc("show", SHOW_SUMMARY)
                .input::<ShowArgs>()
                .positionals(["id"])
                .json_output::<Shown>()
                .bind(TypedRpc::new(verb(server.clone(), show))),
        )
}

/// A verb's call against the live tree of the invoking job's conversation,
/// on behalf of the job's node.
struct Invoked<H: AgentHarness, A> {
    tree: Rc<Tree<H>>,
    server: Rc<AgentServer<H>>,
    caller: NodeId,
    json: bool,
    args: A,
}

/// Binds `run` to the tree and node a call names. A call from a job whose
/// tree is not live fails.
fn verb<H, A, F, Fut>(
    server: Weak<AgentServer<H>>,
    run: F,
) -> impl Fn(Call<A>, RpcPort) -> futures_util::future::LocalBoxFuture<'static, Result<u8, RpcError>>
where
    H: AgentHarness,
    A: 'static,
    F: Fn(Invoked<H, A>, RpcPort) -> Fut + Copy + 'static,
    Fut: std::future::Future<Output = Result<u8, RpcError>> + 'static,
{
    move |call: Call<A>, port: RpcPort| {
        let server = server.clone();
        Box::pin(async move {
            let server = server
                .upgrade()
                .ok_or_else(|| RpcError::Failed("the agent server is gone".into()))?;
            let context = &call.invocation.context;
            let root = NodeId::try_from(context.conversation.as_str())
                .map_err(|error| RpcError::Failed(error.to_string()))?;
            let caller = context
                .caller
                .node()
                .and_then(|node| NodeId::try_from(node).ok())
                .ok_or_else(|| RpcError::Failed("demi agent runs only in an agent's job".into()))?;
            let tree = server
                .tree(&root)
                .ok_or_else(|| RpcError::Failed(format!("the conversation {root} is not open")))?;
            let invoked = Invoked {
                tree,
                server,
                caller,
                json: call.invocation.json,
                args: call.args,
            };
            run(invoked, port).await
        })
    }
}

async fn spawn<H: AgentHarness>(
    call: Invoked<H, SpawnArgs>,
    port: RpcPort,
) -> Result<u8, RpcError> {
    let prompt = trim(&call.args.prompt);
    if prompt.is_empty() {
        return fail(&port, "spawn", "prompt must not be empty").await;
    }
    let input = StartInput::Spawn {
        prompt: prompt.to_owned(),
        profile_name: call.args.profile,
        description: call.args.description.unwrap_or_default(),
        is_spawn_forbidden: call.args.no_subagents.unwrap_or(false),
    };
    let request = call.args.request_id.unwrap_or_else(|| call.server.new_id());
    match call.tree.start(&call.caller, input, request, &port).await {
        Ok(child) => started(&port, call.json, child).await,
        Err(error) => fail(&port, "spawn", &error).await,
    }
}

async fn resume<H: AgentHarness>(
    call: Invoked<H, ResumeArgs>,
    port: RpcPort,
) -> Result<u8, RpcError> {
    let message = trim(&call.args.message);
    if message.is_empty() {
        return fail(&port, "resume", "message must not be empty").await;
    }
    let id = match NodeId::try_from(call.args.id.as_str()) {
        Ok(id) => id,
        Err(error) => return fail(&port, "resume", &error.to_string()).await,
    };
    let input = StartInput::Resume {
        id,
        message: message.to_owned(),
    };
    let request = call.args.request_id.unwrap_or_else(|| call.server.new_id());
    match call.tree.start(&call.caller, input, request, &port).await {
        Ok(child) => started(&port, call.json, child).await,
        Err(error) => fail(&port, "resume", &error).await,
    }
}

async fn send<H: AgentHarness>(call: Invoked<H, SendArgs>, port: RpcPort) -> Result<u8, RpcError> {
    if is_blank(&call.args.message) {
        return fail(&port, "send", "message must not be empty").await;
    }
    let message = trim(&call.args.message).to_owned();
    match call
        .tree
        .send_message(&call.caller, &call.args.id, message)
        .await
    {
        Ok(target) if call.json => {
            json(
                &port,
                &Sent {
                    id: target,
                    accepted: true,
                },
            )
            .await
        }
        Ok(target) => out(&port, format!("sent to {target}\n")).await,
        Err(error) => fail(&port, "send", &error).await,
    }
}

async fn abort<H: AgentHarness>(
    call: Invoked<H, AbortArgs>,
    port: RpcPort,
) -> Result<u8, RpcError> {
    let id = call.args.id;
    let Some(child) = NodeId::try_from(id.as_str())
        .ok()
        .filter(|child| call.tree.is_child_of(child, &call.caller))
    else {
        return fail(
            &port,
            "abort",
            &format!("\"{id}\" is not one of your running children"),
        )
        .await;
    };
    call.tree.abort_child(&child).await;
    if call.json {
        return json(
            &port,
            &Aborted {
                id: child,
                aborted: true,
            },
        )
        .await;
    }
    out(&port, format!("aborted {child}\n")).await
}

async fn list<H: AgentHarness>(call: Invoked<H, ListArgs>, port: RpcPort) -> Result<u8, RpcError> {
    let listing = match call.tree.listing().await {
        Ok(listing) => listing,
        Err(error) => return fail(&port, "list", &error).await,
    };
    if call.json {
        return json(
            &port,
            &Listing {
                tree: listing.entries(&call.caller),
            },
        )
        .await;
    }
    out(&port, format!("{}\n", listing.render(&call.caller))).await
}

async fn show<H: AgentHarness>(call: Invoked<H, ShowArgs>, port: RpcPort) -> Result<u8, RpcError> {
    let id = call.args.id;
    let Some((snapshot, text)) = NodeId::try_from(id.as_str())
        .ok()
        .and_then(|child| call.tree.show(&child))
    else {
        return fail(&port, "show", &format!("no live agent \"{id}\"")).await;
    };
    if call.json {
        return json(&port, &Shown { agent: snapshot }).await;
    }
    out(&port, text).await
}

async fn started(port: &RpcPort, json_output: bool, child: NodeId) -> Result<u8, RpcError> {
    if json_output {
        return json(port, &Started { subagent_id: child }).await;
    }
    out(port, format!("subagentId: {child}\n")).await
}

async fn json(port: &RpcPort, value: &impl Serialize) -> Result<u8, RpcError> {
    let text = serde_json::to_string(value).expect("a command's output serializes");
    out(port, format!("{text}\n")).await
}

async fn out(port: &RpcPort, text: String) -> Result<u8, RpcError> {
    port.stdout(text.into_bytes()).await?;
    Ok(0)
}

/// A verb's failure: `demi agent <verb>: <reason>` on stderr, exit 1.
async fn fail(port: &RpcPort, verb: &str, reason: &str) -> Result<u8, RpcError> {
    port.stderr(format!("demi agent {verb}: {reason}\n").into_bytes())
        .await?;
    Ok(1)
}
